lasagne/layers/noise.py

```diff
@@ -211,3 +219,76 @@ def get_output_for(self, input, deterministic=False, **kwargs):
+        the layer feeding into this layer, or the expected input shape
+    p : float or scalar tensor
+        The probability of setting a value to zero
+    rescale : bool
+        If ``True`` (the default), scale the input by ``1 / (1 - p)`` when
+        dropout is enabled, to keep the expected output mean the same.
+    shared_axes : tuple of int
+        Axes to share the dropout mask over. By default, each value can be
+        dropped individually. ``shared_axes=(0,)`` uses the same mask across
+        the batch. ``shared_axes=(2, 3)`` uses the same mask across the
+        spatial dimensions of 2D feature maps.
+    alpha : float or SELU instance
+        The default values are fixed point solutions to equations (4) and (5)
+        in [1] for zero mean and unit variance input. The analytic expressions
+        for them are given in equation (14) also in [1].
+
+    Notes
+    -----
+    The alpha dropout layer is a regularizer that randomly sets input values to
+    zero and also applies a normalization function to bring the mean to 0
+    and the variance to 1; see [1]_for why this might improve
+    generalization. The behaviour of the layer depends on the ``deterministic``
+    keyword argument passed to :func:`lasagne.layers.get_output`. If ``True``,
+    the layer behaves deterministically, and passes on the input unchanged. If
+    ``False`` or not specified, dropout (and possibly scaling) is enabled.
+    Usually, you would use ``deterministic=False`` at train time and
+    ``deterministic=True`` at test time.
+
+    References
+    ----------
+    .. [1] Klambauer, G., Unterthiner, T., Mayr, A., Hochreiter, S. (2017):
+           Self-Normalizing Neural Networks. arXiv preprint: 1706.02515
+    """
+
+    def __init__(self, incoming, p=0.1, rescale=True, shared_axes=(),
+                 alpha=None, **kwargs):
+        """Class initialization."""
+        super(AlphaDropoutLayer, self).__init__(incoming,
+                                                p=p,
+                                                rescale=rescale,
+                                                shared_axes=shared_axes,
+                                                **kwargs)
+        from lasagne.nonlinearities import SELU
+        if alpha is None:
+            self.alpha = -1.0507009873554804934193349852946 * \
+                         1.6732632423543772848170429916717
+        elif isinstance(alpha, SELU):
+            self.alpha = - alpha.scale * alpha.scale_neg
+        else:
+            self.alpha = alpha
+
+    def get_output_for(self, input, deterministic=False, **kwargs):
+        """Apply alpha dropout."""
+        if deterministic or self.p == 0:
+            return input
+        else:
+            mask = self.apply_dropout(input, const=self.alpha)
+            a = T.pow(self.q + T.square(self.alpha) * self.q * self.p, -0.5)
+            b = -a * self.p * self.alpha
+
+            return T.cast(a * mask + b, input.dtype)
```

Comment:
I believe it returns a float64 so I just want to make sure the dtype of the output is the same as the dtype of the input